formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/schema/ProtoBufSchemaGenerator.kt

```diff
@@ -0,0 +1,509 @@
+@file:OptIn(ExperimentalSerializationApi::class)
+
+package kotlinx.serialization.protobuf.schema
+
+import kotlinx.serialization.*
+import kotlinx.serialization.descriptors.*
+import kotlinx.serialization.protobuf.*
+
+/**
+ * Experimental generator of ProtoBuf schema that is compatible with [serializable][Serializable] Kotlin classes
+ * and data encoded and decoded by [ProtoBuf] format.
+ *
+ * The schema is generated based on provided [SerialDescriptor] and is compatible with proto2 schema definition.
+ * An arbitrary Kotlin class represent much wider object domain than the ProtoBuf specification, thus schema generator
+ * has the following list of restrictions:
+ *
+ *  * Serial name of the class and all its fields should be a valid Proto [identifier](https://developers.google.com/protocol-buffers/docs/reference/proto2-spec)
+ *  * Nullable values are allowed only for Kotlin [nullable][SerialDescriptor.isNullable] types
+ *  * [SerialName] of the type uniquely identifies the proto message type and two or more fields with the same serial name
+ *    are considered to have the same type.
```

Comment:
It is also important that there is no duplicate extracted names in the passed `SerialDescriptor`s, since they are all considered classes with different structures.